### ISSUE TITLE
VF2 graph isomorphism takes very long or is stuck

### ISSUE THREAD

**Issue by gaborcsardi** [igraph/igraph#238]
See the attached two graphs

library(igraph)
g1 <- read.graph("graph_first.txt", dir=FALSE)
g2 <- read.graph("graph_second.txt", dir=FALSE)
graph.isomorphic.vf2(g1, g2)

---

Imported from Launchpad using lp2gh.
- date created: 2012-09-17T22:05:18Z
- owner: gabor.csardi
- assignee: gabor.csardi
- the launchpad url was https://bugs.launchpad.net/bugs/1052213

**Comment by xratsos** [comment]
Dear Developers,

I recently started using igraph package on R for a graph mining task. I run to the same problem discussed here, so I thought to report it again.
The problem occur on a particular case, when both graphs have the same number of vertices(25) and edges(41). The difference is on two edges direction. I let "vf2" run through all the weekend and it did not finished. Using LAD algorithm the results was given almost instantly ("false").

BR,
Chrats

**Comment by szhorvat** [comment]
Of the isomorphism algorithms included in igraph, Bliss seems to be the fastest by far (at least for large or difficult graphs).  Have you considered using Bliss?

Currently Bliss does not support directed graphs.  Once pull request #873 gets merged, it will get support for directed graphs, vertex colored graphs and returning the generators of the automorphism group (which is typically more useful than returning all isomorphisms one by one, like VF2 does).  There really won't be a good reason to use VF2 except for edge coloured graphs and for subgraph isomorphism.

**Comment by xratsos** [comment]
Thanks for the fast respond. 
Surely I will try Bliss when will be available for directed graphs. Currently, I modified my pipeline to use subgraph_isomorphic(LAD algorithm) instead. This lead to a R total crash, because of a comparison of two other graphs. Anyway by set induced parameter to TRUE now it seems will work(I will check the server tomorrow).

**Comment by szhorvat** [comment]
LAD doesn't support directed graphs either, so I'm not sure how you could use it with them ...

**Comment by xratsos** [comment]
I never worked on graphs and isomorphism before, so you are probable right. I read from the documentation ?subgraph_isomorphic:
"pattern	
The smaller graph, it might be directed or undirected. Undirected graphs are treated as directed graphs with mutual edges.
target	
The bigger graph, it might be directed or undirected. Undirected graphs are treated as directed graphs with mutual edges."

and confuse it?

**Comment by xratsos** [comment]
szhorvat could you give me an estimate when the "bliss" algorithm for directed graphs will be available?
BR,
Chrats

**Comment by szhorvat** [comment]
@xratsos Sorry, I cannot answer that.  I am just a user like you, not an igraph developer.  I contributed a minor change to the C version of igraph that updated Bliss to the latest version that supports directed graphs.  It is still under review.  I am hoping that it will be part of the next release, whenever that happens (I wouldn't expect it soon).  I answered in this thread only because I was interested in the topic.

Regarding LAD, you are correct: it does support directed graphs.
